### 5.4 PREDICCIONES PARA EL CONJUNTO DE LA COMPETICIÓN

En este segundo cuaderno, nos limitamos a cargar el archivo simulado de competición, aplicar el modelo previamente guardado (que ya contiene toda la lógica de transformación de preprocesado gracias a la pipeline) y volcar los resultados en un CSV.

In [1]:
import pandas as pd
import numpy as np
import joblib
from sklearn.preprocessing import FunctionTransformer

def transform_pdays(X):
    X_out = X.copy()
    X_out['wascontacted'] = X_out['pdays'].apply(lambda x: 0 if x == -1 else 1)
    
    def bin_pdays(x):
        if x == -1: return "no_contacto"
        elif x < 100: return "0_99"
        elif x < 200: return "100_199"
        elif x < 400: return "200_399"
        else: return "400_plus"
    
    X_out['pdays_binned'] = X_out['pdays'].apply(bin_pdays)
    X_out = X_out.drop(columns=['pdays'])
    return X_out

# Ahora sí podemos cargar el modelo
modelo = joblib.load('modelo_final.joblib')
print("Modelo cargado correctamente.")

# Cargamos los datos de competición
df_competicion = pd.read_pickle(r"C:\Users\claug\OneDrive\Escritorio\Uni\4 de carrera\aprendizaje automático\Practica1AA\AA_PRACTICA1\bank_competition.pkl")
print(f"Shape del dataset de competición: {df_competicion.shape}")

# Corregimos None en job y marital
df_competicion['job']     = df_competicion['job'].replace({None: np.nan})
df_competicion['marital'] = df_competicion['marital'].replace({None: np.nan})

# Generamos predicciones
predicciones = modelo.predict(df_competicion)

print(f"\nDistribución de predicciones:")
print(pd.Series(predicciones).value_counts())

# Guardamos en CSV
df_predicciones = pd.DataFrame(predicciones, columns=['prediccion'])
df_predicciones.to_csv('predicciones.csv', index=False)

print(f"\nPredicciones guardadas en 'predicciones.csv' ({len(predicciones)} filas).")
print(df_predicciones.head(10))

Modelo cargado correctamente.
Shape del dataset de competición: (162, 16)

Distribución de predicciones:
no     91
yes    71
Name: count, dtype: int64

Predicciones guardadas en 'predicciones.csv' (162 filas).
  prediccion
0         no
1         no
2         no
3        yes
4        yes
5         no
6        yes
7         no
8        yes
9         no


### Conclusión: Predicciones de la competición

La inferencia sobre el conjunto de datos de competición se ha ejecutado de forma exitosa y eficiente, confirmando la solidez de la arquitectura diseñada. Al haber encapsulado toda la lógica dentro de una `Pipeline` (incluyendo la transformación de la variable `pdays`), hemos garantizado que las 162 nuevas instancias reciban exactamente el mismo preprocesamiento que los datos de entrenamiento, evitando errores de formato y fugas de datos. 

Es importante destacar que el único paso manual requerido fue la homogeneización de los valores nulos (`None` a `np.nan`) en las variables `job` y `marital`, replicando el paso previo al flujo de la pipeline que ya establecimos en el entorno de desarrollo. 

Las predicciones generadas muestran una distribución coherente (91 predicciones negativas frente a 71 positivas), lo que indica que el modelo SVM ha generalizado correctamente sin colapsar hacia la clase mayoritaria. Finalmente, los resultados han sido exportados correctamente al archivo `predicciones.csv`, cumpliendo con el objetivo de la competición y dejando el modelo validado para su posterior integración en la interfaz de Streamlit.

### COMPROBACION GIT PUSH

In [ ]:
#iwoe`ndwnco`w